flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py

```diff
@@ -180,3 +184,64 @@ def _read_and_validate_dataframe(
             model = RootModel[list[model]]
         data = [r.model_dump() for r in model.model_validate(data).root]
     return pd.DataFrame.from_records(data)
+
+
+@overload
+def _evaled_expression(val: str, *, target_type: Type[int]) -> int: ...
+@overload
+def _evaled_expression(val: Any, *, target_type: Type[int]) -> int | Any: ...
+
+
+@overload
+def _evaled_expression(val: str, *, target_type: Type[float]) -> float: ...
+@overload
+def _evaled_expression(val: Any, *, target_type: Type[float]) -> float | Any: ...
+
+
+def _evaled_expression(val: Any, *, target_type: Type[EE]) -> EE | Any:
+    """
+    Evaluates a string expression to a target numeric type (int or float).
+
+    Args:
+        val: The input value to process.
+        target_type: The type (int or float) to convert the expression to.
+
+    Returns:
+        The value coerced into the target numeric type, or the original value.
+
+    Raises:
+        ValueError: On parsing errors.
+
+    Examples:
+        >>> _evaled_expression("1 + 1", target_type=int)
+        2
+        >>> _evaled_expression("10 / 4", target_type=float)
+        2.5
+        >>> _evaled_expression("10 / 4", target_type=int)
+        # note that result is trucnated; probably undesirable if misused
+        2
+        >>> _evaled_expression(99.5, target_type=float)
+        99.5
+        >>> _evaled_expression(None, target_type=int)
+
+        >>> _evaled_expression("a * b", target_type=float)
+        Traceback (most recent call last):
+            ...
+        ValueError: Can't convert expression to float.
+    """
+    if isinstance(val, target_type):
+        return val
+
+    if isinstance(val, str):
+        try:
+            return target_type(sympy.parsing.sympy_parser.parse_expr(val))
+        except TypeError as e:
+            raise ValueError(f"Can't convert expression to {target_type.__name__}.") from e
```

Comment:
What value do we get out of wrapping the `TypeError` here, the exception doesn't catch syntax errors. i.e.
```
>>> sympy.parsing.sympy_parser.parse_expr("2..0 + 1")
ValueError: Error from parse_expr with transformed code: "Float ('2.' )Float ('.0' )+Integer (1 )"

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
...
    expr = eval(
           ^^^^^
  File "<string>", line 1
    Float ('2.' )Float ('.0' )+Integer (1 )
                 ^^^^^
SyntaxError: invalid syntax
```
Maybe:
```suggestion
        expr = parse_expr(val)
        if not expr.is_Number:
          raise ValueError(f"Cannot convert expression '{expr}' to {target_type}.")
        return target_type(expr)
```
Also open to other ideas. I'm not super familiar with `sympy` so this may not be the best way to do this.